File: frontend/lib/features/products/data/product.dart
class Product {
  const Product({
    required this.id,
    required this.brand,
    required this.productType,
    required this.category,
    required this.stockQuantity,
    required this.ingredients,
  });

  static const brands = [
    'CERAVE',
    'LA_ROCHE_POSAY',
    'SKINCEUTICALS',
    'OBAGI',
    'ZO_SKIN_HEALTH',
    'BIODERMA',
    'AVENE',
  ];
  static const types = [
    'CLEANSER',
    'MOISTURIZER',
    'SERUM',
    'SUNSCREEN',
    'TONER',
    'EXFOLIANT',
    'MASK',
    'RETINOID',
  ];
  static const ingredientsByType = <String, List<String>>{
    'CLEANSER': ['CERAMIDES', 'SALICYLIC_ACID'],
    'MOISTURIZER': ['HYALURONIC_ACID', 'CERAMIDES'],
    'SERUM': ['VITAMIN_C', 'NIACINAMIDE'],
    'SUNSCREEN': ['ZINC_OXIDE'],
    'TONER': ['NIACINAMIDE', 'GLYCOLIC_ACID'],
    'EXFOLIANT': ['GLYCOLIC_ACID', 'SALICYLIC_ACID'],
    'MASK': ['HYALURONIC_ACID', 'NIACINAMIDE'],
    'RETINOID': ['RETINOL'],
  };

  final String id;
  final String brand;
  final String productType;
  final String category;
  final int stockQuantity;
  final List<String> ingredients;

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as String,
      brand: json['brand'] as String,
      productType: json['productType'] as String,
      category: json['category'] as String,
      stockQuantity: json['stockQuantity'] as int,
      ingredients: List<String>.from(json['ingredients'] as List),
    );
  }

  String get brandLabel => label(brand);
  String get typeLabel => label(productType);

  static String label(String value) {
    return value
        .split('_')
        .map((word) => '${word[0]}${word.substring(1).toLowerCase()}')
        .join(' ');
  }
}

class ProductInput {
  const ProductInput({
    required this.brand,
    required this.productType,
    required this.category,
    required this.stockQuantity,
    required this.ingredients,
  });

  final String brand;
  final String productType;
  final String category;
  final int stockQuantity;
  final List<String> ingredients;

  Map<String, dynamic> toJson() => {
    'brand': brand,
    'productType': productType,
    'category': category,
    'stockQuantity': stockQuantity,
    'ingredients': ingredients,
  };
}

class PatientProductRecord {
  const PatientProductRecord({
    required this.id,
    required this.productId,
    required this.brand,
    required this.productType,
    required this.source,
    this.startedOn,
    this.discontinuedOn,
  });

  final String id;
  final String productId;
  final String brand;
  final String productType;
  final String source;
  final String? startedOn;
  final String? discontinuedOn;

  factory PatientProductRecord.fromJson(Map<String, dynamic> json) {
    return PatientProductRecord(
      id: json['id'].toString(),
      productId: json['productId'].toString(),
      brand: json['brand'].toString(),
      productType: json['productType'].toString(),
      source: json['source'].toString(),
      startedOn: json['startedOn']?.toString(),
      discontinuedOn: json['discontinuedOn']?.toString(),
    );
  }
}

File: backend/src/main/java/com/example/backend/entities/Product.java
package com.example.backend.entities;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotNull;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

// A catalogue row is a brand plus a type. No soft delete: history points here.
@Entity
@Table(name = "product")
@Getter
@Setter
@NoArgsConstructor
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @NotNull
    @Column(nullable = false, length = 60)
    private String name;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 60)
    private ProductBrand brand;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(name = "product_type", nullable = false, length = 40)
    private ProductType productType;

    @Column(name = "image_url", length = 2048)
    private String imageUrl;

    @Column(name = "stock_quantity", nullable = false)
    private int stockQuantity;

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.ARRAY)
    @Column(nullable = false, columnDefinition = "text[]")
    private List<Ingredient> ingredients = new ArrayList<>();

    public enum ProductBrand {
        CERAVE,
        LA_ROCHE_POSAY,
        SKINCEUTICALS,
        OBAGI,
        ZO_SKIN_HEALTH,
        BIODERMA,
        AVENE
    }

    public enum ProductType {
        CLEANSER,
        MOISTURIZER,
        SERUM,
        SUNSCREEN,
        TONER,
        EXFOLIANT,
        MASK,
        RETINOID
    }

    public enum Ingredient {
        RETINOL,
        NIACINAMIDE,
        SALICYLIC_ACID,
        HYALURONIC_ACID,
        VITAMIN_C,
        GLYCOLIC_ACID,
        BENZOYL_PEROXIDE,
        AZELAIC_ACID,
        CERAMIDES,
        ZINC_OXIDE
    }
}
